lib/puppet/configurer.rb

```diff
@@ -109,8 +109,17 @@ def convert_catalog(result, duration, options = {})
 
     catalog_conversion_time = thinmark do
       # Will mutate the result and replace all Deferred values with resolved values
-      if options[:convert_for_node] && options[:convert_with_facts]
-        Puppet::Pops::Evaluator::DeferredResolver.resolve_and_replace(options[:convert_for_node], options[:convert_with_facts], result)
+      if facts = options[:convert_with_facts]
+        Puppet::Pops::Evaluator::DeferredResolver.resolve_and_replace(facts, result)
+      else
+        # During a cached_catalog run, facts may not have been collected; even in
+        # that case, we should still attempt to reolve_and_replace deferred
+        # functions and send an empty hash for facts parameter.
```

Comment:
The result would be surprising - if someone actually needs a fact value they would get nil.
So - question why does this happen and should something other than resolving deferred values be done?
This seems like a bad idea that will lead to hard to figure out problems for users.

Depending on why this happens it should either get the current facts, or it should error.

Reply:
I think the solution here is to call `get_facts` in `prepare_and_retrieve_catalog_from_cache`, and that way convert_catalog method won't change much, it will just not check for a node definition, and it won't try to resolve a catalog without facts.